## Using RSem

    rsem-prepare-reference --gtf /path to/gencode.v33.annotation.gtf --star --star-path ~/anaconda3/bin/ /path to/GRCh38.primary_assembly.genome.fa /home/parashar/scratch/name_of_output_file 2> rsem.stderr

## Alignments
bsub -e rsem.e -o rsem.o -n 32 -q regularq rsem-calculate-expression --output-genome-bam --strandedness reverse -p 32 --calc-pme --calc-ci --keep-intermediate-files --append-names --sort-bam-by-coordinate --paired-end --estimate-rspd --star --star-path /home/parashar/anaconda3/bin/ /home/parashar/archive/ananda_rnaseq/trimmomatic/paired/Contr_S14_L004_R1_p.fastq /home/parashar/archive/ananda_rnaseq/trimmomatic/paired/Contr_S14_L004_R2_p.fastq /home/parashar/archive/rnaseq/rsem /home/parashar/scratch/last_rsem_test/Contr_S14_L004 2> /home/parashar/scratch/last_rsem_test/Contr_S14_L004.stderr

Bulk RNA-Seq Analysis
It's desirable to generate all the possible files you can if you have large computation power. Since running STAR in RSEM will require 32 GB RAM and 1 TB swappable disk memory, be wise to choose from the below mentioned output files. I personally prefer to generate all the possible output files at once

To do alignment and calculate read abundance in terms of TPM used the following command 

    rsem-calculate-expression --output-genome-bam --star-output-genome-bam --strandedness reverse -p 32 --calc-pme --calc-ci --keep-intermediate-files --append-names --sort-bam-by-coordinate --paired-end --estimate-rspd --star --star-path ~/anaconda3/bin/ ~/path/Contr1_S15_L004_R1_p.fastq ~/path/Contr1_S15_L004_R2_p.fastq ~/path/to/index/hg38 /path/prefix for output file (eg.Contr1_S15_L004) 2> /path/Contr1_S15_L004.stderr

> --output-genome-bam
        Generate a BAM file, 'sample_name.genome.bam', with alignments
        mapped to genomic coordinates and annotated with their posterior
        probabilities. In addition, RSEM will call samtools (included in
        RSEM package) to sort and index the bam file.
        'sample_name.genome.sorted.bam' and
        'sample_name.genome.sorted.bam.bai' will be generated. (Default:
        off)
        --star-output-genome-bam
        (STAR parameter) Save the BAM file from STAR alignment under genomic
        coordinate to 'sample_name.STAR.genome.bam'. This file is NOT sorted
        by genomic coordinate. In this file, according to STAR's manual,
        'paired ends of an alignment are always adjacent, and multiple
        alignments of a read are adjacent as well'. (Default: off)
        --strandedness <none|forward|reverse>
        This option defines the strandedness of the RNA-Seq reads. It
        recognizes three values: 'none', 'forward', and 'reverse'. 'none'
        refers to non-strand-specific protocols. 'forward' means all
        (upstream) reads are derived from the forward strand. 'reverse'
        means all (upstream) reads are derived from the reverse strand. If
        'forward'/'reverse' is set, the '--norc'/'--nofw' Bowtie/Bowtie 2
        option will also be enabled to avoid aligning reads to the opposite
        strand. For Illumina TruSeq Stranded protocols, please use
        'reverse'. (Default: 'none')
        --paired-end
        Input reads are paired-end reads. (Default: off)
        --calc-pme
        Run RSEM's collapsed Gibbs sampler to calculate posterior mean
        estimates. (Default: off)
        --estimate-rspd
        Set this option if you want to estimate the read start position
        distribution (RSPD) from data. Otherwise, RSEM will use a uniform
        RSPD. (Default: off)
        --calc-ci
        Calculate 95% credibility intervals and posterior mean estimates.
        The credibility level can be changed by setting
        --keep-intermediate-files
        Keep temporary files generated by RSEM. RSEM creates a temporary
        directory, 'sample_name.temp', into which it puts all intermediate
        output files. If this directory already exists, RSEM overwrites all
        files generated by previous RSEM runs inside of it. By default,
        after RSEM finishes, the temporary directory is deleted. Set this
        option to prevent the deletion of this directory and the
        intermediate files inside of it. (Default: off)
        --append-names
        If gene_name/transcript_name is available, append it to the end of
        gene_id/transcript_id (separated by '_') in files
        'sample_name.isoforms.results' and 'sample_name.genes.results'.
        (Default: off)
        --sort-bam-by-coordinate
        Sort RSEM generated transcript and genome BAM files by coordinates
        and build associated indices. (Default: off)



## Rationale of using above parameters

 - --estimate-rspd option should be specified so that RSEM can estimate a read start position distribution (RSPD), which may allow for more
   accurate abundance estimates. So just to take care of biases if any.
   
 - Check with the experimentalists which kit was used for library prepation and which protocol was followed. If a strand-specific protocol is used, the --strand-specific option should be specified. Otherwise, it is assumed that a read has equal probability of coming from the sense or antisense directions. Also, you might end ub getting more reads in Ambiguous category. We used Illumina TrueSeq Standard Protocol which gives majority of the transcripts from the reverse strand.

> In addition to computing ML abundance estimates, RSEM can also use a Bayesian version of its model to produce a PME and 95% CI for the abundance of each gene and isoform. These values are computed by Gibbs sampling (see Methods) and can be obtained by specifying the --calc-ci option. The 95% CIs are valuable for assessing differential expression across samples, particularly for repetitive genes or isoforms because the CIs capture uncertainty due to both random sampling effects and read mapping ambiguity. **We recommend using the CIs in combination with the results of differential expression tools, which currently do not take into account variance from multiread allocation.** The PME values may be used in lieu of the ML estimates as they are very similar, but have the convenient property of generally being contained within the 95% CIs, which is sometimes not the case for small ML estimates.

Keep-intermediate files will help store alignment stats generated by star which otherwise is discarded by RSEM. The Log.final.out file in sample.tmp folder help determine the quality of your data. You can also get those stats from .cnt files. See [here](https://groups.google.com/forum/#!topic/rsem-users/usmPKgsC5LU) and [here](https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt) for reading .cnt files.

## ENCODE3 pipeline parameters used by RSEM
                       " --genomeDir $star_genome_path " .
                       ' --outSAMunmapped Within ' .
                       ' --outFilterType BySJout ' .
                       ' --outSAMattributes NH HI AS NM MD ' .
                       ' --outFilterMultimapNmax 20 ' .
                       ' --outFilterMismatchNmax 999 ' .
                       ' --outFilterMismatchNoverLmax 0.04 ' .
                       ' --alignIntronMin 20 ' .
                       ' --alignIntronMax 1000000 ' .
                       ' --alignMatesGapMax 1000000 ' .
                       ' --alignSJoverhangMin 8 ' .
                       ' --alignSJDBoverhangMin 1 ' .
                       ' --sjdbScore 1 ' .
                       " --runThreadN $nThreads " .
                       ##

                       ## different than ENCODE3 pipeline
                       ## do not allow using shared memory
                       ' --genomeLoad NoSharedMemory ' .
                       ##

                       ## different than ENCODE3 pipeline, which sorts output BAM
                       ## no need to do it here to save time and memory
                       ' --outSAMtype BAM Unsorted ' .
                       ##

                       ## unlike ENCODE3, we don't output bedGraph files

                       ' --quantMode TranscriptomeSAM '.
                       ' --outSAMheaderHD \@HD VN:1.4 SO:unsorted '.

                       ## define output file prefix
                       " --outFileNamePrefix $imdName ";
